src/UI/Buyer/src/app/shared/services/oc-reorder/oc-reorder.service.ts

```diff
@@ -0,0 +1,113 @@
+import { orderReorderResponse } from '@app-buyer/shared/services/oc-reorder/oc-reorder.interface';
+import { Observable, of, forkJoin } from 'rxjs';
+import { flatMap } from 'rxjs/operators';
+import { OcMeService, BuyerProduct, LineItem } from '@ordercloud/angular-sdk';
+import {
+  map as _map,
+  forEach as _forEach,
+  find as _find,
+  differenceBy as _differenceBy,
+} from 'lodash';
+
+@Injectable()
+export class AppReorderService {
+  constructor(
+    private appLineItemService: AppLineItemService,
+    private meService: OcMeService
+  ) {}
+
+  order(orderID: string): Observable<orderReorderResponse> {
+    return this.appLineItemService.listAll(orderID).pipe(
+      flatMap((list) => {
+        let lineItems = of(list.Items); // this sets var into an observable
+        let productIds = _map(list.Items, 'ProductID');
+        let validProducts = this.getValidProducts(productIds);
+        return forkJoin([validProducts, lineItems]);
+      }),
+      flatMap((results) => this.isProductInLiValid(results[0], results[1])),
+      flatMap((results) => this.hasInventory(results))
+    );
+  }
+
+  getValidProducts(
+    productIds: string[],
+    validProducts?: BuyerProduct[]
+  ): Observable<BuyerProduct[]> {
+    validProducts = validProducts || [];
+    let chunk = productIds.splice(0, 25);
+    return this.meService
+      .ListProducts({ filters: { ID: chunk.join('|') } })
+      .pipe(
+        flatMap((productList) => {
+          validProducts = validProducts.concat(productList.Items);
+          if (productIds.length) {
+            return this.getValidProducts(productIds, validProducts);
+          } else {
+            return of(validProducts);
+          }
+        })
+      );
+  }
+
+  isProductInLiValid(
+    products: BuyerProduct[],
+    lineItems: LineItem[]
+  ): Observable<orderReorderResponse> {
+    let validProductIDs = _map(products, 'ID');
+    let validLi: LineItem[] = [];
+    let invalidLi: LineItem[] = [];
+
+    _forEach(lineItems, (li) => {
+      if (validProductIDs.indexOf(li.ProductID) > -1) {
+        let product = _find(products, { ID: li.ProductID });
+        li.Product = product;
+        validLi.push(li);
+      } else {
+        invalidLi.push(li);
+      }
+    });
+    return of({ ValidLi: validLi, InvalidLi: invalidLi });
+  }
+
+  hasInventory(
+    response: orderReorderResponse
+  ): Observable<orderReorderResponse> {
+    // compare new validLi with old validLi and push difference into the new invalid[] + old invalid array.
+    let newOrderResponse: orderReorderResponse;
+    let newValidLi = response.ValidLi.filter(isValidInventory);
+    let newInvalidLi = _differenceBy(response.ValidLi, newValidLi, 'ProductID');
+
+    newInvalidLi = newInvalidLi.concat(response.InvalidLi);
+    newOrderResponse = { ValidLi: newValidLi, InvalidLi: newInvalidLi };
+    console.log('here is the order Response', newOrderResponse);
+    return of(newOrderResponse);
+
+    function isValidInventory(li) {
```

Comment:
This is a style thing that can really improve the readability of your code. Instead of all these nesting ifs, negate the conditions and get rid of the nesting. so It becomes

  ``` function isValidInventory(li) {
      let inventory = li.Product.Inventory;
      if (!hasRestrictedQuantity(li)) {
        return false;
      }

      if (!inventory || !inventory.Enabled) {
        return false;
      }

      if (inventory.OrderCanExceed) {
        return true;
      }

      return inventory.QuantityAvailable >= li.Quantity;
    }